src/coreComponents/physicsSolvers/inducedSeismicity/kernels/RateAndStateKernels.hpp

```diff
@@ -0,0 +1,224 @@
+              StackVariables & stack ) const
+  {
+    /// Solve 2x2 system
+    real64 solution[2] = {0.0, 0.0};
+
+    denseLinearAlgebra::solve< 2 >( stack.jacobian, stack.rhs, solution );
+
+    /// Update variables
+    m_stateVariable[k] -= solution[0];
+    m_slipRate[k]      -= solution[1];
+
+    /// Matteo: debugging tools.
+    // printf("solution[0] = %.10e\n", solution[0]);
+    // printf("solution[1] = %.10e\n", solution[1]);
+
+    // printf("m_stateVariable[%d] = %.10e\n", k, m_stateVariable[k]);
+    // printf("m_slipRate[%d] = %.10e\n", k, m_slipRate[k]);
+  }
+
+  GEOS_HOST_DEVICE
+  std::pair< int, real64 > checkConvergence( StackVariables const & stack,
+                                             real64 const tol ) const
+  {
+    real64 const residualNorm = LvArray::tensorOps::l2Norm< 2 >( stack.rhs );
+    int const converged = residualNorm < tol ? 1 : 0;
+    return std::make_pair( converged, residualNorm );
+  }
+
+private:
+
+  arrayView1d< real64 > const m_slipRate;
+
+  arrayView1d< real64 > const m_stateVariable;
+
+  arrayView1d< real64 const > const m_stateVariable_n;
+
+  arrayView2d< real64 const > const m_traction;
+
+  real64 const m_shearImpedance;
+
+  constitutive::RateAndStateFriction::KernelWrapper m_frictionLaw;
+
+};
+
+
+
+/**
+ * @brief Performs the kernel launch
+ * @tparam POLICY the policy used in the RAJA kernels
+ */
+template< typename POLICY >
+static void
+createAndLaunch( SurfaceElementSubRegion & subRegion,
+                 string const & frictionLawNameKey,
+                 real64 const shearImpedance,
+                 integer const maxNewtonIter,
+                 real64 const time_n,
+                 real64 const dt )
+{
+  GEOS_MARK_FUNCTION;
+
+  GEOS_UNUSED_VAR( time_n );
+
+  string const & frictionaLawName = subRegion.getReference< string >( frictionLawNameKey );
+  constitutive::RateAndStateFriction const & frictionLaw = subRegion.getConstitutiveModel< constitutive::RateAndStateFriction >( frictionaLawName );
+  RateAndStateKernel kernel( subRegion, frictionLaw, shearImpedance );
+
+  // Newton loop (outside of the kernel launch)
+  bool allConverged = false;
+  for( integer iter = 0; iter < maxNewtonIter; iter++ )
+  {
+    RAJA::ReduceMin< parallelDeviceReduce, int > converged( 1 );
+    RAJA::ReduceMax< parallelDeviceReduce, real64 > residualNorm( 0.0 );
+    forAll< POLICY >( subRegion.size(), [=] GEOS_HOST_DEVICE ( localIndex const k )
+    {
+      RateAndStateKernel::StackVariables stack;
+      kernel.setup( k, dt, stack );
+      kernel.solve( k, stack );
+      auto result = kernel.checkConvergence( stack, 1.0e-6 );
+      converged.min( std::get< 0 >( result ) );
+      residualNorm.max( std::get< 1 >( result ) );
+    } );
+
+    real64 const maxResidualNorm = MpiWrapper::max( residualNorm.get() );
+    GEOS_LOG_RANK_0( GEOS_FMT( "-----iter {} : residual = {:.10e} ", iter, maxResidualNorm ) );
+
+    if( converged.get() )
+    {
+      allConverged = true;
```

Comment:
do you need mpi for this one?

Reply:
I think that we can do it without mpi. if one rank fails it should just kill them all